src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java

```diff
@@ -104,34 +106,50 @@ public void put(Collection<SinkRecord> records) throws ConnectException {
     partitionPauser.maybeResumePartitions();
 
     for (SinkRecord record : records) {
+      verifyChangingTopic(record);
       OffsetState offsetState = offsetTracker.addPendingRecord(record);
-
       if (shouldSkipRecord(record)) {
         logTrace("Ignoring {} with null value.", record);
         offsetState.markProcessed();
         reportBadRecord(record, new ConnectException("Cannot write null valued record."));
         continue;
       }
-
       logTrace("Writing {} to Elasticsearch.", record);
-
       tryWriteRecord(record, offsetState);
     }
     partitionPauser.maybePausePartitions();
   }
 
+  /**
+   * Fail fast in cases where topic mutating SMTs are used and connector configuration doesn't
+   * support it.
+   */
+  private void verifyChangingTopic(SinkRecord record) {
+    TopicPartition tp = new TopicPartition(record.topic(), record.kafkaPartition());
+    if (!config.flushSynchronously() && !context.assignment().contains(tp)) {
+      String msg = String.format("Found a topic name '%s' that doesn't match assigned partitions."
+          + " Connector doesn't support topic mutating SMTs", record.topic());
+      throw new ConnectException(msg);
+    }
+  }
+
   @Override
   public Map<TopicPartition, OffsetAndMetadata> preCommit(
-          Map<TopicPartition, OffsetAndMetadata> currentOffsets) {
+      Map<TopicPartition, OffsetAndMetadata> currentOffsets) {
     try {
       // This will just trigger an asynchronous execution of any buffered records
       client.flush();
     } catch (IllegalStateException e) {
       log.debug("Tried to flush data to Elasticsearch, but BulkProcessor is already closed.", e);
     }
-    Map<TopicPartition, OffsetAndMetadata> offsets = offsetTracker.offsets();
-    log.debug("preCommitting offsets {}", offsets);
-    return offsets;
+    if (config.flushSynchronously()) {
```

Comment:
should this logic be moved to offset tracker.offsets() method and it would simplify this code and also SyncOffsetTracker can do the waitForInFlightRequests() inside it?

Reply:
Good idea, moved.